### PR TITLE
Implement union operation on primitive sets. Partially addresses issue #310.

### PR DESCRIPTION
**Pull request by prathasirisha**
Signed-off-by: Sirisha Pratha <sirisha.pratha@bnymellon.com>

### CODE REVIEW

File: eclipse-collections-code-generator/src/main/resources/test/set/mutable/abstractImmutablePrimitiveSetTestCase.stg [hunk 1]
```diff
@@ -263,6 +263,16 @@ public abstract class AbstractImmutable<name>HashSetTestCase extends AbstractImm
         Assert.assertEquals(set.toSet(), set.asLazy().toSet());
         Verify.assertInstanceOf(Lazy<name>Iterable.class, set.asLazy());
     }
+
+    @Test
+    public void union()
```
Comment (donraab) [inline eclipse-collections-code-generator/src/main/resources/test/set/mutable/abstractImmutablePrimitiveSetTestCase.stg:268]: I think we should add tests for both sets being empty, and for either being empty.

File: eclipse-collections-code-generator/src/main/resources/test/set/mutable/abstractPrimitiveSetTestCase.stg [hunk 2]
```diff
@@ -457,6 +457,16 @@ public abstract class Abstract<name>SetTestCase extends AbstractMutable<name>Col
         Verify.assertInstanceOf(Unmodifiable<name>Set.class, set.asUnmodifiable());
         Assert.assertEquals(new Unmodifiable<name>Set(set), set.asUnmodifiable());
     }
+
+    @Test
+    public void union()
```
Comment (donraab) [inline eclipse-collections-code-generator/src/main/resources/test/set/mutable/abstractPrimitiveSetTestCase.stg:462]: I think we should add tests for both sets being empty, and for either being empty.

File: unit-tests/src/test/java/org/eclipse/collections/impl/set/mutable/primitive/AbstractByteSetTestCase.java [hunk 3]
```diff
@@ -388,4 +388,14 @@ public void classIsNonInstantiable()
     {
         Verify.assertClassNonInstantiable(ByteSets.class);
     }
+
+    @Test
+    public void union()
```
Comment (donraab) [inline unit-tests/src/test/java/org/eclipse/collections/impl/set/mutable/primitive/AbstractByteSetTestCase.java:393]: I think we should add tests for both sets being empty, and for either being empty.

File: unit-tests/src/test/java/org/eclipse/collections/impl/set/mutable/primitive/AbstractImmutableByteHashSetTestCase.java [hunk 4]
```diff
@@ -228,4 +228,14 @@ public void toImmutable()
         Assert.assertEquals(1, this.newWith((byte) 1).toImmutable().size());
         Assert.assertEquals(3, this.newWith((byte) 1, (byte) 2, (byte) 3).toImmutable().size());
     }
+
+    @Test
+    public void union()
```
Comment (donraab) [inline unit-tests/src/test/java/org/eclipse/collections/impl/set/mutable/primitive/AbstractImmutableByteHashSetTestCase.java:233]: I think we should add tests for both sets being empty, and for either being empty.

**Review by donraab — CHANGES_REQUESTED** [review]
Hi @prathasirisha, thank you for the contribution. It mostly looks good. I've just requested a few extra test scenarios for empty sets.

File: eclipse-collections-code-generator/src/main/resources/api/set/immutablePrimitiveSet.stg [hunk 5]
```diff
@@ -43,6 +43,16 @@ public interface Immutable<name>Set extends Immutable<name>Collection, <name>Set
     @Override
     \<V> ImmutableSet\<V> collect(<name>ToObjectFunction\<? extends V> function);
 
+    /**
+     * @since 11.0.
+     */
+    @Override
+    default Immutable<name>Set union(<name>Set set)
+    {
+        Mutable<name>Set newSet = this.toSet();
```
Comment (mohrezaei) [inline eclipse-collections-code-generator/src/main/resources/api/set/immutablePrimitiveSet.stg:52]: use the larger of the two sets as starting point.

File: eclipse-collections-code-generator/src/main/resources/api/set/mutablePrimitiveSet.stg [hunk 6]
```diff
@@ -72,6 +72,16 @@ public interface Mutable<name>Set extends Mutable<name>Collection, <name>Set
...
         throw new UnsupportedOperationException("Implement in concrete classes.");
     }
+
+    /**
+     * @since 11.0.
+     */
+    @Override
+    default Mutable<name>Set union(<name>Set set)
+    {
+        Mutable<name>Set newSet = this.toSet();
```
Comment (mohrezaei) [inline eclipse-collections-code-generator/src/main/resources/api/set/mutablePrimitiveSet.stg:82]: use the larger of the two sets as starting point.

**Review by mohrezaei — CHANGES_REQUESTED** [review]

**Comment by nikhilnanivadekar** [comment]
We need another API which takes a target set.

File: eclipse-collections-code-generator/src/main/resources/test/set/mutable/abstractImmutablePrimitiveSetTestCase.stg (hunk 1 above)
Comment (prathasirisha) [reply eclipse-collections-code-generator/src/main/resources/test/set/mutable/abstractImmutablePrimitiveSetTestCase.stg:268]: Added tests for 1. both sets being empty. 2. left (set1) is empty 3. right (set2) is empty. 
Lines: 285-298.

File: eclipse-collections-code-generator/src/main/resources/test/set/mutable/abstractPrimitiveSetTestCase.stg (hunk 2 above)
Comment (prathasirisha) [reply eclipse-collections-code-generator/src/main/resources/test/set/mutable/abstractPrimitiveSetTestCase.stg:462]: Added tests for 1. both sets being empty. 2. left (set1) is empty 3. right (set2) is empty.
Lines: 479-492.

File: unit-tests/src/test/java/org/eclipse/collections/impl/set/mutable/primitive/AbstractByteSetTestCase.java (hunk 3 above)
Comment (prathasirisha) [reply unit-tests/src/test/java/org/eclipse/collections/impl/set/mutable/primitive/AbstractByteSetTestCase.java:393]: Added tests for 1. both sets being empty. 2. left (set1) is empty 3. right (set2) is empty.
Lines: 410-423.

File: unit-tests/src/test/java/org/eclipse/collections/impl/set/mutable/primitive/AbstractImmutableByteHashSetTestCase.java (hunk 4 above)
Comment (prathasirisha) [reply unit-tests/src/test/java/org/eclipse/collections/impl/set/mutable/primitive/AbstractImmutableByteHashSetTestCase.java:233]: Added tests for 1. both sets being empty. 2. left (set1) is empty 3. right (set2) is empty.
Lines: 250-263.

File: eclipse-collections-code-generator/src/main/resources/api/set/immutablePrimitiveSet.stg (hunk 5 above)
Comment (prathasirisha) [reply eclipse-collections-code-generator/src/main/resources/api/set/immutablePrimitiveSet.stg]: Updated.

File: eclipse-collections-code-generator/src/main/resources/api/set/mutablePrimitiveSet.stg (hunk 6 above)
Comment (prathasirisha) [reply eclipse-collections-code-generator/src/main/resources/api/set/mutablePrimitiveSet.stg]: Updated. Thanks for the suggestion!

**Review by donraab — APPROVED** [review]
LGTM.

**Review by mohrezaei — APPROVED** [review]

**Comment by donraab** [comment]
> We need another API which takes a target set.

Hi @nikhilnanivadekar, I think we can add the methods which take target collections in a separate PR.
